### PR TITLE
Issue 16485 - Add trait for testing if a member is static.

### PR DESCRIPTION
**Pull request by sprinkle131313**
Opening this new pull request as @jmdavis seems busy and hasn't replied to pings, credit to him (#4834)

### CODE REVIEW

**Comment by RazvanN7** [comment]
LGTM

**Comment by BBasile** [comment]
nope, it'll fail if the inspection is made in another module without a public protection.

To be clear anything like that shouldn't be proposed until [something like this](https://github.com/BBasile/DIPs/blob/3d5e3f81c541c6e23c69555a230b4d42a7bb6de6/DIPs/DIP8484.md)  is implemented.

If you merge this without the new behavior against protection for __trait() then you'll get a bug report, certainly. Can't say when, but this will happen.

File: std/traits.d [hunk 1]
```diff
@@ -3550,6 +3551,176 @@ enum hasMember(T, string name) = __traits(hasMember, T, name);
...
+ *     member = Name of symbol to test that resides in $(D T).
+ *
+ * Returns:
+ *     $(D true) iff $(D member) is static.
+ */
+template isStaticMember(T, string member)
+    if (__traits(hasMember, T, member))
+{
+    import std.meta : Alias;
+    alias sym = Alias!(__traits(getMember, T, member));
```
Comment (BBasile) [inline std/traits.d]: nope, it'll fail if the inspection is made in another module without a public protection.

To be clear anything like that shouldn't be proposed until [something like this](https://github.com/BBasile/DIPs/blob/3d5e3f81c541c6e23c69555a230b4d42a7bb6de6/DIPs/DIP8484.md)  is implemented.

If you merge this without the new behavior against protection for __trait() then you'll get a bug report, certainly. Can't say when, but this will happen.

Comment (sprinkle131313) [inline std/traits.d]: I don't think it causes that big of a problem that it's worth waiting for a DIP that's still being written. When or even if that DIP is introduced, it won't cause any breakage to existing code and it'll be a small patch to allow private protected symbols for isStaticMember. I don't see there being very many cases where it would need to be made known whether a private member is static in another module that it won't be able to access anyways.

**Comment by sprinkle131313** [comment]
Hmm not sure why dlang-bot isn't linking the issue with bugzilla.

**Comment by wilzbach** [comment]
> Hmm not sure why @dlang-bot isn't linking the issue with bugzilla.

You need to mention it with the git commit message.
See also: https://github.com/dlang-bots/dlang-bot#automated-references

**Comment by jmdavis** [comment]
> Opening this new pull request as @jmdavis seems busy and hasn't replied to pings,

Well, I didn't think that there was really anything left to do on it. The only comment on there had to do with adding "returns" to the docs, which seems inappropriate to me, since it's not a function, but whatever.

> No without the required change to the protection when certain __trait() verbs are used. Which is the case here.

I don't think that this particularly matters for this PR. Yes, the language needs to be fixed, but type introspection in general is broken until the language is fixed, and worst case, using this template just results in compilation errors when you use it on something with inaccessible symbols. Yes, it sucks, but there's no fix for it without fixing the language, and it's already been agreed that we're going to fix the language. So, this template will work on some code for now and fail to compile on other code, and then later, it'll start compiling on that code too.

File: std/traits.d (hunk 1 above)
Comment (BBasile) [inline std/traits.d]: It's not a DIP anymore, it'll be done without following the DIP process. The question is when.

Comment (sprinkle131313) [inline std/traits.d]: see https://github.com/dlang/phobos/pull/5112#issuecomment-279196565
Doesn't look like they are going to be going with that exact DIP implementation, so changes won't be necessary.

**Comment by wilzbach** [comment]
@sprinkle131313 this still needs a changelog entry. Please see the `/changelog` folder for instructions and examples.

**Comment by sprinkle131313** [comment]
The instructions to build the changelog only leads to a command that prints that the feature will be implemented soon.

https://github.com/dlang/phobos/tree/master/changelog
https://github.com/dlang/dlang.org/blob/master/posix.mak#L523

**Comment by wilzbach** [comment]
> The instructions to build the changelog only leads to a command that prints that the feature will be implemented soon.

Yeah nobody has merged this PR yet: https://github.com/dlang/dlang.org/pull/1549
Just copy/paste another entry then, the changelog builder when run  for the next release will pick it up.

File: std/traits.d [hunk 2]
```diff
@@ -3554,6 +3555,176 @@ enum hasMember(T, string name) = __traits(hasMember, T, name);
...
+        __gshared void function() gfp;
+        void function() fpm;
+
+        void delegate() dm;
+        static void delegate() sd;
+
+        void m() {}
+        final void m2() const pure nothrow @nogc @safe {}
+
+        inout(int) iom() inout { return 10; }
```
Comment (andralex) [inline std/traits.d]: some of these are not covered; if they are needed as declarations but never called please don't define

**Comment by jmdavis** [comment]
> Cool, please rename to hasStaticMember for symmetry with hasMember. Thanks!

This isn't testing whether the type has a member by that name that's static. It's testing whether a member that it definitely has is static or not. If it were going to test whether a type had a member and if it were static, then its implementation would need to be changed, and doing that would result in unnecessary redundancy, because the typical way to use this is likely to be when you have a list of members already and are checking which ones are static.

**Comment by andralex** [comment]
Oh, then definitely something is off. If a member known to exist is to be verified, then `isStaticMember` should take by alias, e.g. `isStaticMember!(S.abc)` as opposed to `isStaticMember!(S, "abc")`. In that case, it is clearly a member so the name `isStatic` suffices.

**Comment by andralex** [comment]
> the typical way to use this is likely to be when you have a list of members already and are checking which ones are static

I'm not very sure this is the case, what would be some scenarios?

**Comment by jmdavis** [comment]
> Oh, then definitely something is off. If a member known to exist is to be verified, then isStaticMember should take by alias, e.g. isStaticMember!(S.abc) as opposed to isStaticMember!(S, "abc"). In that case, it is clearly a member so the name isStatic suffices.

I believe that the reason this is the way it is is primarily because `__traits(allMembers, ...)` gives all of the members as strings. Presumably, it could be done with symbols, but that would result in a very different implementation and wouldn't play very well with `__traits(allMembers, ...)` - though in code that was already dealing with symbols, it would be easier to deal with. Both approaches have value.

> I'm not very sure this is the case, what would be some scenarios?

Manu is the one who pushed for this originally, and I don't know exactly what his use case was, but he was specifically looking to test whether a member was static, not whether a type had a member with a particular name that was static.

Personally, I've run into the need for a trait like this when dealing with serialization code (e.g. serializing a struct to JSON based on its members). In that case, you're definitely dealing with a list of members and testing whether they're static rather than trying to see if a type has a specific member that is static.

I would think that most cases where you were looking to simply use a static member variable, you wouldn't care whether it was an enum or a static variable, and you could just check whether you could use the member in the context you wanted to, and if it's a static function you're looking for, `__traits(isStaticFunction, ...)` solves the problem easily enough. It's checking whether a specific member is a static member variable that gets hard. Introspection of a type and looking at all of its members is the only case at the moment that I can think of where you really care that a member is static as opposed to just working with the correct syntax.

The original thread (split for some reason):

http://forum.dlang.org/post/mailman.810.1475501243.2994.digitalmars-d@puremagic.com
http://forum.dlang.org/post/mailman.811.1475502086.2994.digitalmars-d@puremagic.com

**Comment by andralex** [comment]
>I believe that the reason this is the way it is is primarily because `__traits(allMembers, ...)` gives all of the members as strings.

If `__traits(allMembers, ...)` is already used then the effort has already been done; there's no advantage in assuming the member exists already. Right now what happens if the member doesn't exist? Is the response `false` or a compile-time error?

@sprinkle131313 you may want to rethink the design, thanks.

**Comment by jmdavis** [comment]
> If __traits(allMembers, ...) is already used then the effort has already been done;

What effort is already done? All you have from `allMembers` is a list of strings that are the names of members of the type. You know nothing about them except that they're members of that type. The hard part here is correctly testing whether a member is static or not. It's surprisingly hard to get right. And if you have the result of `allMembers`, then this is by far the cleanest approach. Actually getting at the symbol to test gets far more complicated.

> Is the response false or a compile-time error?

It's a compiler error. The template constraint requires that the given string be a name of one of the members of the given type. If you used `allMembers`, then you know that the given string is the name of a member of the given type, and you're fine. If you got the name from elsewhere and need to verify that it's a member, then just use `__traits(hasMember, ..)` or `std.traits.hasMember` before using `isStaticMember`.

**Comment by andralex** [comment]
@jmdavis I'm saying the work has already been done to figure out whether it's a member, so it's difficult to justify it. I'd rather use `hasStaticMember!(Foo, "bar")` than `hasMember!(Foo, "bar") && isStaticMember!(Foo, "bar")` if there's no advantage to the latter.

**Comment by jmdavis** [comment]
Hmmm. So, basically, take the current implementation and move the template constraint to a static if? It would be the same test either way, so that does make sense. It is pretty weird from the standpoint of the original goal, since in that case, you already know that it's a member, but it still works, and since the template constraint was testing it anyway, there's no extra overhead for the original use case, and there's less overhead in the case where you don't know for sure that it's a member. Okay. That makes sense.

File: std/traits.d [hunk 3]
```diff
@@ -44,6 +44,7 @@
  *           $(LREF hasElaborateDestructor)
  *           $(LREF hasIndirections)
  *           $(LREF hasMember)
+ *           $(LREF isStaticMember)
```
Comment (jmdavis) [inline std/traits.d]: This still needs to be changed, and your most recent commit has unrelated changes mixed in.

File: changelog/std-traits-hasStaticMember.dd [hunk 4]
```diff
@@ -0,0 +1,13 @@
+Added `std.traits.hasStaticMember` to check whether a symbol is a static member of a type.
+-------
+import std.traits : hasStaticMember;
+
+struct S
+{
+    static int globalVar;
+    int localVar;
```
Comment (andralex) [inline changelog/std-traits-hasStaticMember.dd]: global and local should be nonstatic and static

File: .vscode/tasks.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=733558
+    // for the documentation about the tasks.json format
+    "version": "0.1.0",
```
Comment (andralex) [inline .vscode/tasks.json]: what's with this?

File: std/traits.d [hunk 6]
```diff
@@ -3554,6 +3555,182 @@ enum hasMember(T, string name) = __traits(hasMember, T, name);
...
+        void delegate() dm;
+        static void delegate() sd;
+
+        void m() {}
+        final void m2() const pure nothrow @nogc @safe;
+
+        inout(int) iom() inout { return 10; }
+        static inout(int) iosf(inout int x);
+
+        @property int p() { return 10; }
```
Comment (wilzbach) [inline std/traits.d]: Don't define a body when you don't call it as it will lead to coverage errors ;-)

File: changelog/std-traits-hasStaticMember.dd [hunk 7]
```diff
@@ -0,0 +1,13 @@
+Added `std.traits.hasStaticMember` to check whether a symbol is a static member of a type.
```
Comment (wilzbach) [inline changelog/std-traits-hasStaticMember.dd]: @andralex while the current changelog parser is very forgiving, we should probably make a decision on the format:
- (A) title (can be multiple lines) - empty line - description
- (B) title (one line) - empty line (optional) - description

(A) is implemented as https://github.com/dlang/tools/pull/220
(B) is how our parser works currently.
Personally I am quite happy with closing the PR and agreeing that a title can't be more than one line (doesn't make much sense anyways).

Comment (andralex) [inline changelog/std-traits-hasStaticMember.dd]: I'll let @CyberShadow decide, thanks!

Comment (CyberShadow) [inline changelog/std-traits-hasStaticMember.dd]: I don't see any deciding arguments one way or another.

I think enforcing that the second line is blank would be good to avoid accidental ambiguity, and would follow formatting for git commit messages.

Comment (andralex) [inline changelog/std-traits-hasStaticMember.dd]: Suggestion: start with a more restrictive policy then relax it as experience develops. The other way is much more difficult, as we've all learned the hard way :).

File: std/traits.d (hunk 6 above)
Comment (sprinkle131313) [inline std/traits.d]: Those ones caused link errors if they aren't there.

File: .vscode/tasks.json (hunk 5 above)
Comment (sprinkle131313) [inline .vscode/tasks.json]: My mistake, i hate that ctrl+enter commits everything even if it wasn't staged. I keep doing it by accident.

Comment (wilzbach) [inline .vscode/tasks.json]: How about a new PR that adds these files to the `.gitignore`, s.t. you don't need to worry about it in the future?

File: changelog/std-traits-hasStaticMember.dd (hunk 7 above)
Comment (wilzbach) [inline changelog/std-traits-hasStaticMember.dd]: >I think enforcing that the second line is blank would be good to avoid accidental ambiguity,
>Suggestion: start with a more restrictive policy then relax it as experience develops.

@sprinkle131313 as we seem to have reached a consensus here, would you be so kind to insert an empty line after the title?

Comment (andralex) [inline changelog/std-traits-hasStaticMember.dd]: @wilzbach how do we use that nice relatively new feature of github allowing others to amend someone's pull request?

Comment (CyberShadow) [inline changelog/std-traits-hasStaticMember.dd]: The PR author needs to opt in, which it doesn't look like they've done here.

Comment (wilzbach) [inline changelog/std-traits-hasStaticMember.dd]: https://wiki.dlang.org/Guidelines_for_maintainers#Write_access_to_PRs

It's enabled here:

image

Anyways we put this into the reminder message of the DLang-Bot:
https://github.com/dlang-bots/dlang-bot/pull/44

Comment (CyberShadow) [inline changelog/std-traits-hasStaticMember.dd]: > It's enabled here:

Oh, oops, I was looking in the wrong place.
